script/publish_public_data_json

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
+
+"""
+Export query result data as JSON to a publicly accessible bucket.
+
+Data of the query is exported if "json_export" is set in
+the corresponding metadata file.
+"""
+
+from argparse import ArgumentParser
+import json
+import os
+import sys
+import re
+
+from google.cloud import storage
+from google.cloud import bigquery
+
+# sys.path needs to be modified to enable package imports from parent
+# and sibling directories. Also see:
+# https://stackoverflow.com/questions/6323860/sibling-package-imports/23542795#23542795
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+from bigquery_etl.parse_metadata import Metadata  # noqa E402
+
+
+DEFAULT_BUCKET = "ascholtz-dev-test"  # todo
+DEFAULT_API_VERSION = "v1"
+METADATA_FILE = "metadata.yaml"
+SUBMISSION_DATE_RE = re.compile(r"^submission_date:DATE:(\d\d\d\d-\d\d-\d\d)$")
+QUERY_FILE_RE = re.compile(r"^.*/([a-zA-Z0-9_]+)/([a-zA-Z0-9_]+)_(v[0-9]+)/query\.sql$")
+
+parser = ArgumentParser(description=__doc__)
+parser.add_argument(
+    "--target-bucket",
+    "--target_bucket",
+    default=DEFAULT_BUCKET,
+    help="GCP bucket JSON data is exported to.",
+)
+parser.add_argument(
+    "--project_id",
+    default="moz-fx-data-derived-datasets",
+    help="Run query in the target project",
+)
+parser.add_argument(
+    "--api_version",
+    "--api-version",
+    default=DEFAULT_API_VERSION,
+    help="API version data is published under in the storage bucket",
+)
+parser.add_argument("--parameter", help="Query parameters, such as submission_date")
+parser.add_argument(
+    "--query-file", "--query_file", help="File path to query to be executed"
+)
+
+
+def publish_table_as_json(
+    bucket, client, api_version, dataset, table, version, result_table, date=None
+):
+    """Export the `result_table` data as JSON to Cloud Storage."""
+    # "*" makes sure that files larger than 1GB get split up into multiple JSON files
+
+    prefix = f"api/{api_version}/tables/{dataset}/{table}/{version}/files/"
+
+    if date is not None:
+        # if date exists, then query is incremental and newest results are exported
+        prefix += f"{date}/"
+
+    table_ref = client.get_table(result_table)
+
+    print("jsonn")
+
+    job_config = bigquery.ExtractJobConfig()
+    job_config.destination_format = "NEWLINE_DELIMITED_JSON"
+    destination_uri = f"gs://{bucket}/" + prefix + "*.json"
+    extract_job = client.extract_table(
+        table_ref, destination_uri, location="US", job_config=job_config
+    )
+    extract_job.result()
+
+    gcp_convert_ndjson_to_json(bucket, prefix)
+
+
+def gcp_convert_ndjson_to_json(bucket, gcp_path):
+    """Converts ndjson files on GCP to json files."""
+    storage_client = storage.Client()
+
+    blobs = storage_client.list_blobs(bucket, prefix=gcp_path)
+
+    for blob in blobs:
+        content = blob.download_as_string().decode("utf-8").strip()
```

Comment:
The GCS API has everything we need to stream the contents in and stream contents out to a new object without pulling down the entire object to memory, so I'm hoping we can avoid needing high-mem instances.

I have not done this in the python API, but it should be possible. In particular, we should be able to replace `\n` in the input stream with `,\n`, add a `[` character at the beginning and end of the stream.

Reply:
Alright, I'll look into it.

Reply:
Thinking about this a bit more, wouldn't it also be easier for consumers to read ndjson instead of json? That way streaming of json objects would be easier?

Reply:
I understand that thought process. This format was agreed on across across the data team because we expect consumers will not be data engineers but researchers and data scientists, who are not usually exposed to ndjson.